### PR TITLE
Feat: new multisig operations subscription service

### LINKED ISSUE #2476: Tech research. How to subscribe to multisig storage
Is it possible to subscribe to `multisig_multisigs` pallet storage?
If yes, then rework how multisig operations are being retreived.

### PR DESCRIPTION
**Pull request by Asmadek**
* create multisig pallet
* create multisig domain
* create multisig operations feature

### CODE REVIEW

File: src/renderer/shared/pallet/multisig/schema.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { type z } from 'zod';
+
+import { pjsSchema } from '@/shared/polkadotjs-schemas';
+
+export type Timepoint = z.infer<typeof timepoint>;
+export const timepoint = pjsSchema.object({
```
Comment (johnthecat) [inline src/renderer/shared/pallet/multisig/schema.ts]: Can you rename it to `multisigTimepoint`?

File: src/renderer/shared/pallet/multisig/storage.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  }
+
+  return query;
+};
+
+export const storage = {
+  /**
+   * Get list of all multisig operations for given account
+   */
+  multisigs(api: ApiPromise, accountId: AccountId) {
```
Comment (johnthecat) [inline src/renderer/shared/pallet/multisig/storage.ts:27]: Maybe it's better to support multiple accounts? Is it possible in this particular request?

File: src/renderer/shared/pallet/multisig/storage.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+          'key',
+          pjsSchema
+            .storageKey(pjsSchema.accountId, pjsSchema.u8Array)
+            .transform(([accountId, callHash]) => ({ accountId, callHash })),
+        ],
+        ['multisig', pjsSchema.optional(multisig)],
+      ),
+    );
+
+    return substrateRpcPool.call(() => getQuery(api, 'multisigs').entries(accountId)).then(a => schema.parse(a));
```
Comment (johnthecat) [inline src/renderer/shared/pallet/multisig/storage.ts]:
```suggestion
    return substrateRpcPool.call(() => getQuery(api, 'multisigs').entries(accountId)).then(schema.parse);
```

File: src/renderer/domains/multisig/model/multisigs/types.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import { type BN } from '@polkadot/util';
+
+import { type CallData, type CallHash, type HexString } from '@/shared/core';
+import { type AccountId } from '@/shared/polkadotjs-schemas';
+
+export type Timepoint = {
+  height: number;
```
Comment (johnthecat) [inline src/renderer/domains/multisig/model/multisigs/types.ts]: `BlockHeight` from `polkadotjs-schemas`?

File: src/renderer/domains/multisig/model/multisigs/types.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  index: number;
+};
+
+export type MultisigEvent = {
+  accountId: AccountId;
+  status: 'approved' | 'rejected';
+  blockCreated: number;
```
Comment (johnthecat) [inline src/renderer/domains/multisig/model/multisigs/types.ts]: `BlockHeight` from `polkadotjs-schemas`?

File: src/renderer/domains/multisig/model/multisigs/types.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+};
+
+export type Multisig = {
+  status: 'pending' | 'cancelled' | 'executed' | 'error';
+  accountId: AccountId;
+  callHash: CallHash;
+  callData?: CallData;
+  deposit: BN;
+  depositor: AccountId;
+  blockCreated: number;
```
Comment (johnthecat) [inline src/renderer/domains/multisig/model/multisigs/types.ts]: `BlockHeight` from `polkadotjs-schemas`?

File: src/renderer/features/multisig-operations/model/status.ts [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    chains: networkModel.$chains,
+    wallet: walletModel.$activeWallet,
+  },
+  ({ apis, chains, wallet }) => {
+    if (nullable(wallet)) return null;
+
+    const input = [];
+
+    for (const account of wallet.accounts) {
+      if ('chainId' in account && account.chainId) {
```
Comment (sokolova-an) [inline src/renderer/features/multisig-operations/model/status.ts]: `if (account.chainId)` should work

File: src/renderer/features/multisig-operations/model/status.ts [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+export const multisigOperationsFeatureStatus = createFeature({
+  name: 'multisigOperations',
+  input: $input,
+});
+
+multisigOperationsFeatureStatus.start();
+
+sample({
+  clock: walletModel.$activeWallet,
+  filter: (wallet) => walletUtils.isMultisig(wallet),
```
Comment (sokolova-an) [inline src/renderer/features/multisig-operations/model/status.ts]:
```suggestion
  filter: walletUtils.isMultisig,
```

File: src/renderer/domains/collectives/model/referendum/model.ts [hunk 9]
```diff
@@ -26,19 +26,19 @@ const { pending, subscribe, unsubscribe, received, fulfilled } = createDataSubsc
 >({
   initial: $list,
   fn: ({ api, palletType }, callback) => {
-    let currectAbortController = new AbortController();
+    let currentAbortController = new AbortController();
```
Comment (tuul-wq) [inline src/renderer/domains/collectives/model/referendum/model.ts]: This name seems a bit cleaner
```suggestion
    let abortController = new AbortController();
```

File: src/renderer/domains/multisig/model/multisigs/model.ts [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+    let newStore = {};
+
+    for (const { api, accountId } of params) {
+      const chainId = api.genesisHash.toHex();
+
+      newStore = setNestedValue(
+        store,
+        chainId,
+        accountId,
+        multisigOperationService.mergeMultisigOperations(store[chainId]?.[accountId] || [], result[chainId] || []),
```
Comment (tuul-wq) [inline src/renderer/domains/multisig/model/multisigs/model.ts]: Let's divide this into variables for readability

File: src/renderer/domains/multisig/model/multisigs/model.ts [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+});
+
+const { subscribe, unsubscribe } = createDataSubscription<
+  Store,
+  RequestParams[],
+  { callHash: CallHash; multisigId: AccountId; chainId: ChainId } & MultisigEvent
+>({
+  initial: $multisigOperations,
+  fn: (params, callback) => {
+    const unsubscribe: Promise<void | (() => void)>[] = [];
```
Comment (tuul-wq) [inline src/renderer/domains/multisig/model/multisigs/model.ts]: `() => void` can be replaced by `VoidFunction`, it's up to you

File: src/renderer/domains/multisig/model/multisigs/model.ts [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+        polkadotjsHelpers
+          .subscribeSystemEvents(
+            {
+              api,
+              section: `multisig`,
+              // TODO: add NewMultisig event
+              methods: ['MultisigApproval', 'MultisigExecuted', 'MultisigCancelled'],
+            },
+            event => {
+              const accountIdId = 0;
```
Comment (tuul-wq) [inline src/renderer/domains/multisig/model/multisigs/model.ts]: Looks like pure constants, can be moved out of the function

File: src/renderer/domains/multisig/model/multisigs/model.ts [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+  RequestParams[],
+  { callHash: CallHash; multisigId: AccountId; chainId: ChainId } & MultisigEvent
+>({
+  initial: $multisigOperations,
+  fn: (params, callback) => {
+    const unsubscribe: Promise<void | (() => void)>[] = [];
+
+    for (const { accountId, api } of params) {
+      unsubscribe.push(
+        polkadotjsHelpers
```
Comment (tuul-wq) [inline src/renderer/domains/multisig/model/multisigs/model.ts]: Let's reduce nesting

File: src/renderer/domains/multisig/model/multisigs/model.ts [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+
+    return () => {
+      Promise.all(unsubscribe);
+    };
+  },
+  map: (store, { result: { callHash, multisigId, chainId, ...event } }) => {
+    const newStore = cloneDeep(store);
+
+    newStore[chainId] = newStore[chainId] || {};
+    newStore[chainId][multisigId] = newStore[chainId][multisigId] || [];
```
Comment (tuul-wq) [inline src/renderer/domains/multisig/model/multisigs/model.ts]: No sure about this kind of assignment. If `newStore[chainId]` is big, then reassigning leads to worse performance
A bulletproof solution 😉 
```javascript
if (!newStore[chainId]) {
  newStore[chainId] = {};
}
```

File: src/renderer/domains/multisig/model/multisigs/service.ts [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+  isSameMultisig: (a: Multisig, b: Multisig) => {
+    const isSameCallHash = a.callHash === b.callHash;
+    const isSameTimepoint = a.blockCreated === b.blockCreated && a.indexCreated === b.indexCreated;
+    const isSameAccount = a.accountId === b.accountId;
+
+    return isSameCallHash && isSameTimepoint && isSameAccount;
+  },
+
+  isSameEvent: (a: MultisigEvent, b: MultisigEvent) => {
+    return a.accountId === b.accountId && a.blockCreated === b.blockCreated && a.indexCreated === b.indexCreated;
```
Comment (tuul-wq) [inline src/renderer/domains/multisig/model/multisigs/service.ts]: Let's divide into variables like you did in `isSameMultisig`

File: src/renderer/domains/multisig/model/multisigs/service.ts [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+  },
+
+  mergeEvents: (oldEvents: MultisigEvent[], newEvents: MultisigEvent[]) => {
+    return [...oldEvents, ...newEvents.filter(e => !oldEvents.find(o => multisigOperationService.isSameEvent(o, e)))];
```
Comment (tuul-wq) [inline src/renderer/domains/multisig/model/multisigs/service.ts]: Same suggestion here

File: src/renderer/features/multisig-operations/model/status.ts [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+
+sample({
+  clock: debounce(networkModel.$apis, 2000),
+  source: networkModel.$chains,
+  fn: (chains, apis) => {
+    const multisigChains = Object.values(chains)
+      .filter((chain) => networkUtils.isMultisigSupported(chain.options))
+      .map((c) => c.chainId);
+
+    return multisigChains.filter((chainId) => apis[chainId]).join(',');
```
Comment (tuul-wq) [inline src/renderer/features/multisig-operations/model/status.ts]: We can include this filter into previous one 
```javascript
networkUtils.isMultisigSupported(chain.options) && apis[chainId]
```
And use variables for readability

File: src/renderer/features/multisig-operations/model/status.ts [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+import { nullable } from '@/shared/lib/utils';
+import { type AccountId } from '@/shared/polkadotjs-schemas';
+import { networkModel, networkUtils } from '@/entities/network';
+import { walletModel, walletUtils } from '@/entities/wallet';
+
+const $trigger = createStore<string>('');
+const $debouncedApis = createStore<Record<ChainId, ApiPromise>>({});
+
+sample({
+  clock: debounce(networkModel.$apis, 2000),
```
Comment (tuul-wq) [inline src/renderer/features/multisig-operations/model/status.ts:16]: 👍

File: src/renderer/shared/pallet/multisig/storage.ts (hunk 2 above)
Comment (Asmadek) [reply src/renderer/shared/pallet/multisig/storage.ts:27]: `entries` supports only same params as original query, we can't use multiple accounts with it

**Review by johnthecat — APPROVED** [review]

**Pull request merged**